sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -557,6 +578,27 @@ case class Sin(child: Expression) extends UnaryMathExpression(math.sin, "SIN")
   """)
 case class Sinh(child: Expression) extends UnaryMathExpression(math.sinh, "SINH")
 
+@ExpressionDescription(
+  usage = """
+    _FUNC_(expr) - Returns inverse hyperbolic sine of `expr`.
+  """,
+  arguments = """
+    Arguments:
+      * expr - hyperbolic angle
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(0);
+       0.0
+  """)
+case class Asinh(child: Expression)
+  extends UnaryMathExpression((x: Double) => math.log(x + math.sqrt(x * x + 1.0)), "ASINH") {
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    defineCodeGen(ctx, ev, c =>
+      s"${ev.value} = java.lang.Math.log($c + java.lang.Math.sqrt($c * $c + 1.0));")
```

Comment:
what if the input is negative?

Reply:
It will return NaN, it is the same behavior as for existing sinh implementation.

Reply:
mmh..what do other DBs do in this case? I checked the implementation of `FastMath` and it handles negatives...

Reply:
thanks for reminding me of FastMath lib. Maybe we can simply use that lib here, without this hand-crafted calculation?

Reply:
Also I just tried the code with several negative values, i got correct return values.

Reply:
> Also I just tried the code with several negative values, i got correct return values.

Sorry what did you try?

Anyway, if we don't have a dependency on it, I am not sure it is a good idea to add a dependency for only a couple of functions...

Reply:
took a closer look at asinh\atanh implementation in FastMath. the negative handling inside is to apply calculation optimization on abs(double a) and then apply proper sign at the end of the function to return, which should produce the same value as the original formula i'm using here (https://en.wikipedia.org/wiki/Inverse_hyperbolic_functions#Definitions_in_terms_of_logarithms). But FastMath runs faster than the original functions though.